README
================================================================================
=                  Bluelog - Bluetooth Scanner and Logger                      =
=                       Tom Nardi (MS3FGX@gmail.com)                           =
================================================================================

Bluelog is a Bluetooth scanner designed to tell you how many discoverable
devices there are in an area as quickly as possible. It is intended to be used
as a site survey tool, identifying the number of possible Bluetooth targets
there are in the surrounding environment.

I wrote Bluelog because there didn't appear to be any Bluetooth scanner for
Linux that would simply log discoverable devices without also doing a whole
bunch of other stuff I didn't really need.

Of course, in the end I managed to expand Bluelog into doing a whole bunch of
other stuff I didn't really need. But at least I still kept it optional, so you
can always stick with the basics if that's all you want.

--------------------------------------------------------------------------------
- Compatibility                                                                -
--------------------------------------------------------------------------------

Bluelog has been written with portability and efficiency in mind, so it's able
to run on a number of systems and hardware platforms. Basically, as long as the
device can run (and get results from) "hcitool scan", and you can compile
software for it, there is a good chance Bluelog can run on it.

In addition to running on all major Linux distributions, Bluelog has been used
successfully on Chrome OS (running on the CR-48 netbook), the Pwnie Express Pwn
Plug, and OpenWRT devices. Bluelog is architecture portable, and has been
tested on x86, ARM, MIPS, and PowerPC processors.

As of version 1.0.2, Bluelog is included in BackTrack Linux and the development
version of OpenWRT. Build scripts for Bluelog are also available via the Arch
Linux AUR community repository. Bluelog is also included on the Pwn Pad and Pwn
Plug penetration testing devices from Pwnie Express.

--------------------------------------------------------------------------------
- Requirements                                                                 -
--------------------------------------------------------------------------------

Bluelog's only software requirement is BlueZ, and naturally your system must
have at least one Bluetooth adapter connected to it. Any Bluetooth adapter
supported under Linux should work with Bluelog.

On Debian based systems like Ubuntu, you may need to install the "dev" version
of the BlueZ package, which will probably be called something like
"libbluetooth-dev" or "bluez-libs-devel". Some newer distributions break up
the libraries from the main packages, so simply having the BlueZ package
installed might not be enough.

--------------------------------------------------------------------------------
- Installation                                                                 -
--------------------------------------------------------------------------------

To compile Bluelog, simply run the command "make" in this directory. If you get
any message about missing libraries, use your distribution's package manager to
install them. This will let you run Bluelog from the current directory for a
quick test, but not install it to the system.

To actually install Bluelog, use the command "make install". Since this will
install Bluelog to the system directories, you will need to run this as root or
through sudo.

When and if you want to remove Bluelog, you would use the command "make
uninstall". As with "install", this makes changes to the system directories and
will therefore require root permissions.

In addition, if you are upgrading from a previous version, there is also the
command "make upgrade". This will remove any older versions of Bluelog, compile
the new version, and finally install it to the system. As you might have
guessed, this also requires root permissions.

Finally, if you plan on using "Bluelog Live", check out the README.LIVE file
for information on the extra steps required.

--------------------------------------------------------------------------------
- Usage                                                                        -
--------------------------------------------------------------------------------

As of the current version, Bluelog supports the following options:

-i <hci interface or MAC>
    This option tells Bluelog which Bluetooth device you want to scan with.
You can use either the HCI device name (like hci2) or the MAC of the local
adapter. As a bonus, if you give a device which doesn't exist, Bluelog will
fall back on autodetection to find a working device. 

-o <filename>
    This is the (optional) filename of the log file to write. The default
filename has the format of "bluelog-YYYY-MM-DD-HHMM.log", located in the
current directory.

-r <retries>
   This option sets how many attempts Bluelog should make to resolve a device
name. For various reasons (poor link, busy device, etc, etc), devices won't
always respond to a name request in a timely manner. This fills the logs or
Live display with un-named devices which look, frankly, uncool. By default,
Bluelog will make 3 attempts to resolve a device name, using this option you
can set that count to either be lower (faster, but less accurate), or higher
(slower, but possibly more accurate).

-a <minutes>
   This option enables "amnesia mode", which causes Bluelog to forget it has
seen a particular device after a set amount of time, given here as minutes.
When Bluelog encounters a device it has forgotten through this option, it
will print it to the logs again as if it was the first time it has been
seen, and the time found will be updated.

A value of 0 here will cause Bluelog to continuously record a device as
fast as possible (actual speed depends on platform Bluelog is running on).

-w <seconds>
    This is an experimental option that allows adjusting how long Bluelog
instructs BlueZ to scan for. Generally speaking, shorter scan times allow
Bluelog to process the incoming data faster, but requires more processing
power. Longer scan times should theoretically work better on lower end
hardware.

I would recommend not touching this setting unless you know what you're
doing. The current accepted range is 4 to 30 seconds.

-b
   This option will set the log format so that the resulting data is suitable
for upload to ronin's Bluetooth Profiling Project (BlueProPro). This overrides
most other logging options and disables Bluelog Live. For more information on
this project, and the additional steps required to submit your data for
inclusion, visit: www.hackfromacave.com

-c
   This option toggles writing the raw device class to the log file. Enabling
this option disables the -f option. Default is disabled.

-d
   This option will daemonize Bluelog so that it runs in the background. You
will still see the boilerplate and startup messages, but after that you will
no longer see any info from Bluelog in the terminal.

-e
   Use this option to toggle CRC32 MAC address encoding. With this option
enabled, the discovered MAC addresses will never be logged to disk, rather,
each device will have a unique ID generated for it. This prevents privacy
concerns during activities such as Bluetooth traffic monitoring. Default is
disabled.

-f
   This option takes the device class and interprets it into a more human
friendly format. It will tell you what class the device is and also what it's
core capabilities are. For example, the class "0x7a020c" would appear as:
"Smart Phone,(Net Capture Obex Audio Phone)". Enabling this option disables the
-c option. Default is disabled.

-k
   When running an instance of Bluelog in daemon mode, the -k option can be
used to kill it.

-l
   This option switches Bluelog over to Live mode, which uses an automatically
updated web page to show results rather than the console and regular log files.
See README.LIVE for more details.

-n
    Use this option to toggle displaying device names for discovered devices.
Finding the device name takes extra time during scanning, and occasionally
fails. Therefore by not resolving device names, Bluelog can scan faster and
more accurately. Default is disabled.

-m
    This option, if enabled in the build, performs hardware manufacturer
lookups of discovered devices via the MAC OUI. The hardware manufacturer will
be logged in the standard log file, as well as Bluelog Live. The manufacturer
database needs to be installed for this function to work, which makes it
prohibitively large for some platforms (such as OpenWRT).

-q
   Turn off nonessential terminal output. In normal mode this means you will
only see the start time of the scan and the message indicating proper
shutdown. When used with daemon mode (-d), there will be no terminal output
at all. The only exception to this option are critical errors, for obvious
reasons.

-s
    Use this option to toggle syslog only mode. In this mode Bluelog will not
write its normal log file, and instead write only to the system log file
(/var/log/syslog). This mode is especially useful when combined with a network
aware syslog daemon, which can be used to add rudimentary central logging to
multiple Bluelog nodes.

All normal log options apply in syslog mode. Syslog mode cannot be combined
with BlueProPro or Live modes. Default is disabled.

-t
    Use this option to toggle displaying timestamps for both the start and end
of the scan and each new device found in the log file. Default is disabled.

-v
    Use this option to toggle displaying found devices on the console. Verbose
output will also contain device class information and timestamps. Default is
disabled.

-x
   Use this option to toggle MAC address obfuscation. With this option
enabled, Bluelog will display the manufacturer portion of each discovered
MAC, but block out the device specific identifier. Default is disabled.

--------------------------------------------------------------------------------
- Basic Scanning                                                               -
--------------------------------------------------------------------------------

There isn't a whole lot to say about this one. Start up Bluelog with the
appropriate options, and then just walk/drive around the area and see what you
come up with.

For your first scan, try something simple like:

$ bluelog -vtn

This will turn on verbose output, timestamps, device names, and output
to the default log file. There are a number of other options which you can use
to customize the level of logging Bluelog will do, but most people will
probably be happy with just the time, MAC, and device name.

--------------------------------------------------------------------------------
- Bluelog Live                                                                 -
--------------------------------------------------------------------------------

"Bluelog Live" is an alternate interface for Bluelog. Instead of outputting
discovered devices to the console, or writing them to the sparse log file,
Bluelog will create a web page with the results that you can serve up for
anyone who cares to look.

For more information on Bluelog Live, please see the README.LIVE file.

--------------------------------------------------------------------------------
- Daemon Mode                                                                  -
--------------------------------------------------------------------------------

Running Bluelog with the -d option will start it in daemon mode, which puts it
into the background. This mode is especially useful when running Bluelog Live.

Only one instance of Bluelog can run at a time, so if you attempt to start
Bluelog (in either daemon or interactive mode) while it is already running in
daemon mode, you will be prompted to kill the existing process. You can use the
-k option to kill a running Bluelog process, or simply find the process with
ps and kill it manually.

It is worth noting that enabling daemon mode also overrides some other options,
such as verbose mode (since there is no terminal output once Bluelog goes into
the background).

--------------------------------------------------------------------------------
- Linux Kernel 3.0.x Bug                                                       -
--------------------------------------------------------------------------------

Please note that there is a fatal bug in the Linux 3.0.x kernel series which
completely breaks Bluetooth scanning. This bug effects ALL software which uses
Bluetooth, not just scanners like Bluelog. It will even hinder normal usage of
Bluetooth, since it's pretty hard to pair a new device to your computer if you
can't even scan for one.

If your distribution is running Linux 3.0.x (such as Ubuntu 11.10 or Mint 12),
you'll need to upgrade to AT LEAST the 3.1 series to use Bluetooth properly.

You can read more about this bug on the linux-kernel mailing list:

http://marc.info/?l=linux-kernel&m=131629118406044

--------------------------------------------------------------------------------
- Packaging                                                                    -
--------------------------------------------------------------------------------

Bluelog's Makefile supports using environment variables to modify the
installation so it can be more easily built into a binary package for your
distribution. You can pass compiler options and the installation directory to
the Makefile like so:

$ make CFLAGS="-O2 -march=i486 -mtune=i686"
gcc -O2 -march=i486 -mtune=i686 -lbluetooth bluelog.c -o bluelog
$ mkdir ./pkg
$ make install DESTDIR=./pkg
mkdir -p ./pkg/usr/bin/
mkdir -p ./pkg/usr/share/doc/bluelog-0.9.6/
mkdir -p ./pkg/var/lib/bluelog/
cp bluelog ./pkg/usr/bin/
cp ChangeLog COPYING README TODO ./pkg/usr/share/doc/bluelog-0.9.6/
cp -a --no-preserve=ownership www/* ./pkg/var/lib/bluelog/

This allows you to easily script the build process and create a package using
the package creation software for your distribution of choice.

If you are able to build packages for your distribution, please let me know and
we can work something out for hosting them on the site.

--------------------------------------------------------------------------------
- Acknowledgements                                                             -
--------------------------------------------------------------------------------

The initial code for Bluelog was based on sample code included in the book
"Bluetooth Essentials for Programmers", by Albert Huang. This is a very
informative book, and helps a lot if you are looking to get into BlueZ
programming. It almost makes up for the terrible documentation from the BlueZ
project.

The website for this book is located at: http://www.btessentials.com/

Bluelog also implements device class parsing code from "Inquisition", written
by Michael John Wensley and released under the GPLv2.

You can read more about "Inquisition" from his site: http://www.wensley.org.uk/

Bluelog implements a modified version of the CRC hashing functions from
"CRC Tester" by Sven Reifegerste.

Sven's page about CRC encoding: http://www.zorc.breitbandkatze.de/crc.html

The device cache rewrite took inspiration, if not literal code, from
"SpoofTooph" by .ronin. You can check out "SpoofTooph" and .ronin's other
projects on his site: http://www.hackfromacave.com/

Bluelog's UDP functions are based on code submitted by Ian Macdonald
(ianmac51@googlemail.com).

Bluelog also uses some code inspired by functions from pidfile.c by
Martin Schulze.

The font used for Bluelog Live's logo is "Electric Boots" by Jakob Fischer.
You can see a collection of his fonts at: http://www.pizzadude.dk/

The OpenWRT version of Bluelog would not have been possible without the work
of Gary Bonner and the logistical support of Joshua Hurst and Dean Nielson.
Thanks also to Stephen Walker, who maintains the official OpenWRT packages for
Bluelog, and has been an invaluable source of information on the platform.

Thanks to Jonas "onny" Heinrich for maintaining the Arch Linux build script
for Bluelog. Read about his projects at: http://www.project-insanity.org/

Thanks to Dave Porcello, Jonathan Cran, and the entire Pwnie Express team for
their support and assistance on the Pwn Plug optimized build of Bluelog.

Thanks to Paolo Valleri and the Integreen project for patches to Bluelog
developed during the construction of their traffic monitoring network for
Bolzano, Italy. Project site: http://www.integreen-life.bz.it/

Thanks to Teresa Brooks for adding the DEFCON Privacy Village theme for
Bluelog Live.

Thanks to all users who have taken the time to contact me with comments and
suggestions about Bluelog, which keeps pushing me in the right direction.

Finally, a special thanks to those who have donated Bluetooth devices to me for
calibration purposes. Writing a Bluetooth scanner without any devices to scan
is rather difficult, so the hardware has been very valuable to me.

--------------------------------------------------------------------------------
- License                                                                      -
--------------------------------------------------------------------------------

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License version 2 as published by the Free
Software Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

For details, see the file "COPYING" in the source directory.

--------------------------------------------------------------------------------
- More Info                                                                    -
--------------------------------------------------------------------------------

For more information and updates, please see www.digifail.com


ChangeLog
07/19/17:
	Switch to using OUI list from linuxnet.ca

04/21/16:
	Do basic HTML check before displaying device name in Live, blocks "Tribeca Exploit"

04/17/16:
	Theme error message about CGI running as root
	Add DEFCON Privacy Village theme for Bluelog Live, thanks Teresa

03/19/15:
	Fix CSS/CGI paths when using Live in non-root dir (experimental)

04/16/14:
	Remove commas from manufacturer names in OUI list

02/20/14:
	Add option to send UDP server disconnect message
	Update example bluelog.conf
	Add simple UDP server script

02/07/14:
	Terminate lines in config file with semicolon
	Use yes/no in config file instead of 1/0

02/07/14:
	Handle spaces in config file entries (Thanks, Ian)

01/31/14:
	Add prefix option, print node name to UDP messages
	Store local MAC in config struct
	Cleanly close UDP socket
	Update docs/man/help

01/25/14:
	Merge in WIP UDP mode code

01/21/14:
	Fix manufacturer mode not working with encode or obfuscate
	Fix Amnesia mode not working with encode or obfuscate
	Better separate user-visible and internal copies of discovered MAC

01/19/14:
	Fix "Quiet" mode not working from config file
	Enable setting BT interface in config file

01/13/14:
	Fix name resolution always being on.

12/23/13:
	Gracefully handle missing value in config file

11/09/13:
	Created default bluelog.conf file
	Makefile updated to create /etc/bluelog
	oui.txt moved to /etc/bluelog

11/08/13:
	Merged in experimental config file support.
	Not all options are yet supported in config file, but seems to be
	working. Any testing would be appreciated.

11/01/13: Released as Version 1.1.2
	Fix compiler warning for stable release. I'll break that later.

10/29/13:
	Minor Bluelog theme touchups
	Add basic theming info to README.LIVE

10/20/13:
	Add documentation for manufacturer lookups "-m"
	Higher resolution Bluelog logo, scale down in CSS for themes
	Added GitHub README.md

10/14/13:
	Add support for additional CSS elements in Live UI
	Add link to GitHub page in Live "About"
	Merge in libmackerel 1.1
	LCARS looking much better

10/13/13:
	New default theme for Bluelog Live, WIP
	Improvements to LCARS theme

04/11/13:
	Initial work on Pwn Pad target:
	No Live, OUI enabled, changed default log location, UI changes
	Cache size reduced for Pwn Plug and Pwn Pad

04/04/13:
	Added LCARS theme for Bluelog Live. Problem?

04/03/13:
	Fix Makefile after removing doc files

04/01/13:
	Removed README.BAKTRK, redundant info
	Removed FUTURE, most major items accomplished

03/31/13: 
	Makefile now downloads OUI list during build

03/30/13: Released as Version 1.1.1
	Added "standalone" target to Makefile, kind of hacky
	Removed SSI and Perl Bluelog Live modes, CGI is the future!
	Bluelog Live can now display device manufacturer if launched with "-m"
	Updated README.LIVE (needs improvement)

03/27/13:
	Merged in libmackerel
	Merged in MACLIST from haraldscan
	Experimental manufacturer lookups (currently x86 only)
	Configuration options broken out into config.h

02/24/13:
	Allow for friendly class names in verbose mode, thanks Dean

01/31/13:
	EXPERIMENTAL: Added -e option to encode MACs with CRC32
	Updated MAN page
	Updated README

01/24/13:
	Improved memory management, thanks Paolo

12/28/12:
	Improved file cleanup, less idiotic

12/25/12: Released as Version 1.1.0 (Come All Ye Faithful Edition)
	Removed CGI mobile page, not ready for stable release
	Updated README

12/17/12:
	Only log device discovery to syslog if enabled with -s
	Added -w option to adjust scan window
	Reordered MAN page

12/13/12:
	Makefile improvements, added "nolive" to install without Live Mode
	Added -q option to suppress all terminal output
	Updated documentation

11/13/12:
	Changed install directory to /usr/share/bluelog on x86
	Docs updated accordingly

10/23/12:
	Restructure help output
	Clarifications to README.LIVE

09/24/12:
	Updated license language in README

09/20/12:
	Improved error message for BlueZ lockup

08/01/12:
	Added mobile.html
	Added mobile.cgi, lame hack to fix issue
	CGI: Work on mobile support
	Live: Theme font fixes

07/24/12:
	Modified Makefile to ignore all hidden files when building tarball

06/22/12:
	Removed "test" directory, obviously not supposed to be there
	Modified Makefile to exclude git on release

06/18/12: Released as Version 1.0.4

06/15/12:
	Experimental: Adjust scan time based on platform.

06/10/12:	
	PWN/WRT: New error handler. Bluelog will accept up to 5 BlueZ errors.
	CGI: Cleaned up Status pane
	Updates to README.LIVE
	Slight changes to class DB

06/09/12:
	CGI: New Status pane in Live view
	Live: Theme cleanups

06/07/12:
	CGI: Allow version/debug before root user bail out
	Removed README.WRT

06/05/12:
	Added "pwnplug" target to Makefile
	CGI: Theme in-browser error messages
	Live: Fix paths in about/contact
	Live: "parse log file"? What was that supposed to mean?
	Live: Bumped up about/contact font to 12pt

06/01/12:
	WRT: Only WRT theme included in bluelog-live, package shrunk by 24 KB

05/29/12:
	CGI: Added -h option, gives brief info
	CGI: Added -v option, gives module version
	CGI: Added -d option, gives debug into about module configuration

05/21/12:
	WRT: Fix CSS path in CGI

05/21/12:
	CGI: Refuse to run as root. Seems like a good idea.
	CGI: Fixed cgi-bin path
	Live: Removed unused image.

05/18/12:
	Removed start button from Perl and CGI UI, need to figure out fix

05/16/12:
	Initial version of livelog.cgi, a C CGI module for Bluelog Live
	Largely re-written HTML to support CGI module, index.html
	README.LIVE updated with livelog.cgi info
	Added OpenWRT Bluelog Live theme (openwrt.css)

05/15/12:
	Updated livelog.pl, can now create full Live mode HTML
	README.LIVE updated with livelog.pl info
	Fixes to HTML headers in Live

05/14/12: Released as Version 1.0.3
	Switched from "tac" to "sed" in Bluelog Live, WRT compat improvement
	livelog.pl updated, now creates complete Bluelog Live HTML page

05/03/12:
	New: Added -s option to log only to syslog (experimental)

livelog.c
/*
 *  livelog - Simple CGI module for Bluelog Live mode
 * 
 *  Written by Tom Nardi (MS3FGX@gmail.com), released under the GPLv2.
 *  For more information, see: www.digifail.com
 */

#include <stdio.h>
#include <stdlib.h>
#include <syslog.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>
#include <sys/types.h>

// Defines
#define APPNAME "livelog.cgi"
#define VERSION "1.1"
#define MAXNUM 4096
#define MAXLINE 500
#define INFO "/tmp/info.txt"
#define LOG "/tmp/live.log"
#define PID_FILE "/tmp/bluelog.pid"

// Conditionals
#if defined OPENWRT || PWNPLUG
#define CSSPREFIX "/bluelog/"
#else
#define CSSPREFIX "../"
#endif

// Global variables
int mobile;

// Found device struct
struct btdev
{
	char time[20];
	char addr[18];
	char name[248];
	char class[248];
	char hwinfo[248];
};

// Global variables
// Device file
FILE *logfile;
// Status file
FILE *infofile; 
// Init device cache, index
struct btdev dev_cache[MAXNUM];
int device_index = 0;

// Experimental, print all HTML from CGI module
void print_html(char *CSSFILE)
{	
	// File info
	printf("<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional//EN\"\n"
		"\"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\">\n");
	
	// Boilerplate
	printf("<!--This file created by %s (v%s) by MS3FGX-->\n", APPNAME, VERSION);	
	
	printf("<html>\n"
		"<head>\n"
		"<title>Bluelog Live</title>\n"
		"<meta http-equiv=\"Content-Type\" content=\"text/html; charset=iso-8859-1\" >\n");
	

	
	// CSS file prefix
	printf("<link href=\"%s%s\" type=\"text/css\" rel=\"stylesheet\" /></head><body>\n", CSSPREFIX,CSSFILE);

	printf("<link rel=\"icon\" type=\"image/png\" href=\"/images/favicon.png\" />\n"
		"<link rel=\"shortcut icon\" href=\"/images/favicon.png\" />\n"
		"<META HTTP-EQUIV=\"refresh\" CONTENT=\"20\">\n"
		"</head>\n"
		"<body>\n"
		"<div id=\"container\">\n"
		"<div id=\"header\">\n"
		"<span style=\"position:relative;left:15px;top:45px\">\n"
		"<a href=\"http://www.digifail.com/\" target=\"_blank\" title=\"DigiFAIL\"><img src=\"/images/digifail_logo.png\" border=\"0\"></a>\n"
		"</span>\n"
		"</div>\n");

	printf("<div class=\"footer\">\n"
		"<h4>\n"
		"<a href=\"about.html\" onClick=\"window.open('about.html','About','toolbar=no,location=no,directories=no,status=no,menubar=no,scrollbars=no,resizable=no,width=500,height=320,left=50,top=50'); return false;\">About</a>\n"
		"&nbsp;&nbsp;\n"
		"<a href=\"contact.html\" onClick=\"window.open('contact.html','Contact','toolbar=no,location=no,directories=no,status=no,menubar=no,scrollbars=no,resizable=no,width=500,height=510,left=50,top=50'); return false;\">Contact</a>\n"
		"</h4>\n"
		"</div>\n");
}

void print_header(char *CSSFILE)
{		
	// CGI header
	printf("Content-type: text/html\n\n");
	// Boilerplate
	printf("<!--This file created with %s (v%s) by MS3FGX-->\n", APPNAME, VERSION);
	// HTML head
	printf("<html><head><link href=\"%s%s\" type=\"text/css\" rel=\"stylesheet\" /></head><body>\n", CSSPREFIX,CSSFILE);
}

void setup_table()
{
	// Table setup
	puts("<table border=\"1\" cellpadding=\"5\" cellspacing=\"5\" width=\"100%\">\n");
	
	if (!mobile)
	{
		puts("<tr>\n"\
		"<th>Time Discovered</th>\n"\
		"<th>MAC Address</th>\n"\
		"<th>Device Name</th>\n"\
		"<th>Device Class</th>\n"\
		"<th>Hardware Info</th>\n"\
		"</tr>\n");
	}
	else
	{
		puts("<tr>\n"\
		"<th>Time</th>\n"\
		"<th>MAC</th>\n"\
		"<th>Name</th>\n"\
		"<th>Class</th>\n"\
		"</tr>\n");
	}
}

void print_info()
{
	// Line length
	char line [128];
	
	while (fgets(line,sizeof line,infofile))
		fputs(line,stdout);
}

void read_log()
{
	char line [MAXLINE];	
	while (fgets(line,sizeof line,logfile))
	{
		// Copy line to buffer
		char* line_buffer = strdup(line);
	
		// Populate cache entry from file
		strcpy(dev_cache[device_index].time,strsep(&line_buffer, ","));	
		strcpy(dev_cache[device_index].addr,strsep(&line_buffer, ","));	
		strcpy(dev_cache[device_index].name,strsep(&line_buffer, ","));
		strcpy(dev_cache[device_index].class,strsep(&line_buffer, ","));
		strcpy(dev_cache[device_index].hwinfo,strsep(&line_buffer, "\n"));
		
        // Increment index
		device_index++;
	}
}

void print_devices()
{
	while (device_index > 0)
	{	
		// Write out valid table HTML
		printf("<tr>");
		printf("<td>%s</td>",dev_cache[device_index-1].time);		
		printf("<td>%s</td>",dev_cache[device_index-1].addr);	
		
		// Before writing out device name to HTML, do some VERY basic sanitization (not safe, just to block obvious stuff)		
		if ((strstr(dev_cache[device_index-1].name, ">")) || (strstr(dev_cache[device_index-1].name, "</")))
			strcpy(dev_cache[device_index-1].name, "<p style='color:red;'>Blocked Possible Exploit</p>");

		printf("<td>%s</td>",dev_cache[device_index-1].name);	
		printf("<td>%s</td>",dev_cache[device_index-1].class);	
		
		if(!mobile)
			printf("<td>%s</td>",dev_cache[device_index-1].hwinfo);	
		
		printf("</tr>\n");
		device_index--;
	}
}

int read_pid()
{
	// Any error will return 0
	FILE *pid_file;
	int pid;

	if (!(pid_file=fopen(PID_FILE,"r")))
		return 0;
		
	if (fscanf(pid_file,"%d", &pid) < 0)
		pid = 0;

	fclose(pid_file);	
	return pid;
}

void TopBar()
{
	// Discovered devices display
	puts("<div id=\"content\">\n");
	printf("Discovered Devices: %i</div>\n",device_index);
	// Close up info pane
	puts("</div>\n");	
}

void SideBar()
{	
	// Start sidebar
	puts("<div id=\"sidebar\">\n");
	
	// Aux pane
	puts("<div id=\"sideobject\">\n"\
	"<div id=\"auxbox1\">\n"\
	"</div>\n"\
	"</div>\n");
	
	// Start info pane
	puts("<div id=\"sideobject\">\n"\
	"<div id=\"boxheader\">Info</div>\n"\
	"<div id=\"sidebox\">\n"\
	"<div id=\"sidecontent\">");
	
	// Populate sidebar from files
	print_info();
	
	// Close info pane
	puts("</div>\n"\
	"</div>\n"\
	"</div>\n");
	
	// Status pane
	puts("<div id=\"sideobject\">\n"\
	"<div id=\"boxheader\">Status</div>\n"\
	"<div id=\"sidebox\">\n"\
	"<div id=\"sidecontent\">");
	
	// Print current PID status
	puts("<div class=\"sideitem\">");
	puts("Bluelog: ");	
	if (read_pid() != 0)
		puts("<span style=\"color: #00ff00;\"><b>RUNNING</b></span></div>\n");
	else
		puts("<span style=\"color: #ff0000;\"><b>STOPPED</b></span></div>\n");	
	
	// Print number of discovered devices
	puts("<div class=\"sideitem\">");
	printf("Discovered Devices: %i</div>\n",device_index);
	
	// Close status pane
	puts("</div>\n"\
	"</div>\n"\
	"</div>\n");
	
	// Aux pane
	puts("<div id=\"sideobject\">\n"\
	"<div id=\"auxbox2\">\n"\
	"</div>\n"\
	"</div>\n");
	
	// Close sidebar
	puts("</div>\n");
}

void shut_down(void)
{
	// Close files
	fclose(infofile);
	fclose(logfile);
	exit(1);
}

static void help(void)
{
	printf("%s (v%s) by MS3FGX\n", APPNAME, VERSION);
	printf("----------------------------------------------------------------\n");
	printf("This is a CGI module used to create a live webpage of Bluelog\n"
		"results. This module simply parses the log files created by\n"	
		"Bluelog, it has no scanning or logging capability of its own.\n");
	printf("\n");
	printf("For more information, see www.digifail.com\n");
	printf("\n");
	printf("Options:\n"
		"\t-m              Mobile format\n"
		"\t-h              Display help\n"
		"\t-d              Print Debug Info\n"
		"\t-v              Print Version Info\n"		
		"\n");
}

static void debug(void)
{
printf("System Variables\n");
printf("--------------------------\n");
printf("Module Version: %s\n", VERSION);
printf("Max Devices: %i\n",MAXNUM);
printf("Max Line Length: %i\n",MAXLINE);
printf("\n");
printf("File Locations\n");
printf("--------------------------\n");
printf("Info File: %s\n",INFO);
printf("Log File: %s\n", LOG);
printf("CSS Prefix: %s\n",CSSPREFIX);
}

static struct option main_options[] = {
	{ "help", 0, 0, 'h' },
	{ "debug", 0, 0, 'd' },
	{ "version", 0, 0, 'v' },
	{ "mobile", 0, 0, 'm' },
	{ 0, 0, 0, 0 }
};
 
int main(int argc, char *argv[])
{		
	// Variables	
	// Default CSS
	char CSSFILE[12]="style.css";

	// Pointers to filenames
	char *infofilename = INFO;
	char *logfilename = LOG;
	
	// Handle arguments
	int opt;
	while ((opt=getopt_long(argc, argv, "dvhm", main_options, NULL)) != EOF)
	{
		switch (opt)
		{
		case 'h':
			help();
			exit(0);
		case 'd':
			debug();
			exit(0);
		case 'v':
			printf("%s (v%s) by MS3FGX\n", APPNAME, VERSION);
			exit(0);
			break;
		case 'm':
			strcpy(CSSFILE, "mobile.css");
			mobile = 1;
			break;
		default:
			printf("Unknown option.\n");
			exit(0);
		}
	}
	
	// Bail out if we are root, except on WRT
	#ifndef OPENWRT
	if(getuid() == 0)
	{
		syslog(LOG_ERR,"CGI module refusing to run as root!");
		
		// Make sure error message is themed
		print_header(CSSFILE);
		puts("<div id=\"content\">");
		puts("Server attempting to run CGI module as root, bailing out!");
		puts("<p>");
		puts("Check your web server configuration and try again.");
		puts("</body></html>");
		exit(1);
	}
	#endif

	// Read in environment variable
	// Comment this to fix compiler warning until ready to implement
	//char* env_string;
	//env_string=getenv("QUERY_STRING");

	// Print HTML head
	print_header(CSSFILE);
	//print_html(CSSFILE);
	
	// Start container div
	if (!mobile)
		puts("<div id=\"container\">\n");

	// Open files
	if ((infofile = fopen(infofilename, "r")) == NULL)
	{
		syslog(LOG_ERR,"Error while opening %s!",infofilename);
		puts("<div id=\"content\">");
		printf("Error while opening %s!\n",infofilename);
		puts("</body></html>");
		exit(1);
	}
	
	if ((logfile = fopen(logfilename, "r")) == NULL)
	{
		syslog(LOG_ERR,"Error while opening %s!",logfilename);
		puts("<div id=\"content\">");
		printf("Error while opening %s!\n",logfilename);
		puts("</body></html>");
		exit(1);
	}
	
	// Draw sidebar\topbar
	read_log();
	if (!mobile)
		SideBar();
	else
		TopBar();
	
	// Content window
	puts("<div id=\"content\">");
	
	if (device_index > 0)
	{
		// Print results
		setup_table();
		print_devices();
		puts("</table>\n");
	}
	
	// Close content, body, and HTML
	puts("</div></body></html>");
	
	// Close files and exit
	shut_down();
	return 0;
}

